[CLIPPER-117] List apps and get app info requests served by mgmtfrontend API

nishadsingh1 opened the pull request:
The management frontend API now has 2 new GET endpoints:
1. /get_application (returns a JSON object with full info about the requested app)
2. /get_applications (returns a list of names of all apps or, if provided `verbose`==True, a list of full JSON objects of all apps)

The returned JSON objects are in the same format that is requested by /add_application. The intent is that there will now be a uniform externally-facing app info format that is distinct from our redis storage format. See `management_frontend_tests.cpp` for examples.

Python management library now uses these endpoints. Their return types have changed slightly; they now return application objects in the same format requested by /add_application.

The task corresponding to exposing application utilities through the API: https://clipper.atlassian.net/browse/CLIPPER-117

If this approach seems reasonable, we can replicate it for models and containers.

Commit:
Management frontend API now serves GET requests for application info

Commit:
Management frontend API now serves GET requests for application info

AmplabJenkins commented:
Can one of the admins verify this patch?

Commit:
Cleanup

Commit:
Cleanup

Corey-Zumar commented on management/clipper_manager.py:
```diff
@@ -280,26 +280,26 @@ def register_application(self,
-            else:
-                print("Clipper has no applications registered")
-                return ""
+            If `verbose` == False, the list contains the apps' names.
```
Let's list `verbose` under a separate `Parameters` documentation header instead of `Returns`. See `register_applications()` for reference.

nishadsingh1 replied:
Cool, makes sense.

Corey-Zumar commented on src/libclipper/include/clipper/json_util.hpp:
```diff
@@ -65,6 +74,20 @@ rapidjson::Value& check_kv_type_and_return(rapidjson::Value& d,
 }
 
+/* Check for matching types else throw exception */
+rapidjson::Value& check_kv_type_is_bool_and_return(rapidjson::Value& d,
```
The reason for defining this method for the `bool` case is that `check_kv_type_and_return` doesn't support checking multiple types, yet we are required to match against `rapidjson::kFalseType` and `rapidjson::kTrueType` for booleans. We could consider modifying `check_kv_type_and_return` to take a vector of types, but  this solution seems fine - just throwing the idea out there.

nishadsingh1 replied:
Yep, that could also definitely work. Both of these solutions seem pretty hacky; I'll look into a cleaner fix.

Corey-Zumar commented on src/libclipper/include/clipper/json_util.hpp:
```diff
@@ -214,6 +246,12 @@ rapidjson::Value& get_object(rapidjson::Value& d, const char* key_name) {
   return object;
 }
 
+rapidjson::Value& get_array(rapidjson::Value& d, const char* key_name) {
```
This is currently unused. Is there a purpose for including it in addition to our type-specific `get_array` methods (`get_int_array`, etc)?

nishadsingh1 replied:
Yep, it's unnecessary – forgot to clean it up.

Corey-Zumar commented on src/libclipper/include/clipper/json_util.hpp:
```diff
@@ -335,6 +373,82 @@ void add_object(rapidjson::Document& d, const char* key_name,
+}
+
+/* Sets `d` to an array containing info from `candidate_models_redis_format`*/
+void set_candidate_models_doc(std::string& candidate_models_redis_format,
```
We should test this method because it depends on a specific redis format that may change in the future. Consider creating a mock list of candidate models adhering to the current format as in `JsonUtilTests#TestParseCandidateModels` and invoking set_candidate_models_doc on this mock.

Corey-Zumar commented on src/libclipper/include/clipper/json_util.hpp:
```diff
@@ -335,6 +373,82 @@ void add_object(rapidjson::Document& d, const char* key_name,
+}
+
+/* Sets `d` to an object containing reformatted info from `app_info`*/
+void set_app_info_doc(std::unordered_map<std::string, std::string>& app_info,
```
Write a unit test for this method as well

Corey-Zumar commented on src/libclipper/src/redis.cpp:
```diff
@@ -34,6 +34,17 @@ std::unordered_map<string, string> parse_redis_map(
 
+std::vector<string> parse_redis_vector(
+        const std::vector<string>& redis_data) {
+  return redis_data;
```
Unintentional return statement? Please document this method once it's fixed or remove it if it's unnecessary.

nishadsingh1 replied:
Thanks for the catch; the whole method is unnecessary.

Corey-Zumar commented on src/management/src/management_frontend.hpp:
```diff
@@ -44,6 +51,7 @@ const std::string ADD_MODEL = ADMIN_PATH + "/add_model$";
 const std::string GET_METRICS = ADMIN_PATH + "/metrics$";
 const std::string GET_SELECTION_STATE = ADMIN_PATH + "/get_state$";
 const std::string GET_APPLICATIONS = ADMIN_PATH + "/get_applications$";
```
For clarity, I think we should rename `get_applications` to `get_all_applications`. This creates a more intuitive distinction between the endpoint that retrieves all applications and the endpoint that retrieves a specific application based on user-provided parameters.

nishadsingh1 replied:
I like that :)

dcrankshaw commented:
jenkins test this please

Corey-Zumar commented on src/management/src/management_frontend.hpp:
```diff
@@ -259,6 +317,67 @@ class RequestHandler {
+   *  "verbose" := bool
+   * }
+   *
+   * \return Returns a JSON string that encodes a list with info about
```
As in `clipper_manager.py`, move `verbose` under a separate `params` documentation section

nishadsingh1 replied:
Are you sure about this? The other methods in management_frontend.hpp have their desired json input schema shown in the same way that it is here.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/135/
Test FAILed.

Corey-Zumar commented on src/management/src/management_frontend_tests.cpp:
```diff
@@ -89,6 +89,270 @@ TEST_F(ManagementFrontendTest, TestAddApplicationMalformedJson) {
+
+  // The JSON provided for adding the app contains the same name-value
+  // attribute pairs that should be returned by `get_application`.
+  ASSERT_EQ(request_doc, response_doc);
```
This is fine for now. I think it's possible that in the future we will specify JSON flags for an `add_application` call that should not be returned to the user by `get_application`.

Corey-Zumar commented on management/clipper_manager.py:
```diff
@@ -312,25 +312,19 @@ def get_app_info(self, name):
+            Returns a dictionary with the specified application's info. This
+            will contain the attribute name-value pairs that were provided to
+            `register_application`. If no application with name `name` is
+            registered with Clipper, None is returned.
```
This method does not return `None` if no application with name `name` is registered. Rather, invoking `get_app_info("nonexistent_name")` returns the following JSON dictionary:

```
{
  "name": "nonexistent_name"
}
```

nishadsingh1 replied:
Great, thank you for the catch; I was manually adding the name back in within the endpoint handling function without checking if redis actually returned a valid app object.

Corey-Zumar reviewed (changes requested):
Looks good in general - noted a few minor errors and commented with requests for documentation & formatting nits. 

When you're done making changes, run `/bin/format_code.sh` in order to pass code format checks necessary for merge.

dcrankshaw committed:
[CLIPPER-113] Implement model active version tracking and updates (#115)

* changed applications to take unversioned models

* added set_model_version to clipper_manager.py

* don't try to start containers for externally managed models

* code compiles

* unit tests pass

* updated tutorial and removed selection policy section

* addressed code review comments

* added checks for duplicate add_model and add_application calls

* test for successful model publish before deploying model data

* Fixed unittests

* fixed double cleanup call in unittests.sh causing non-zero exit code

* Updated tutorial to reflect tf model as new version of sklearn model

Commit:
[CLIPPER-139] Simplified python predict function deployment (#111)

* Simplified python predict function deployment

* Undid unintentional changes, added description of Anaconda env expectation

* Added accidentally removed import statements

* clean up fixed a few bugs

* fixed serialized predict function file name

* Made requested changes

* Flush output manually, do not run containers in unbuffered mode

dcrankshaw committed:
moved dockerfiles into same directory as build context (#120)

dcrankshaw committed:
[CLIPPER-105] Refactor selection policy to support single-model applications with default output (#89)

* draft of the selection policy rewrite

* adapted query processor to use new selection policy interface

* minor cleanup

* added default output to application table

* things are compiling but there's a linker error from header-only json_utils

* unit tests passing

* addressed review comments and fixed end to end bench

* fixed unittests

* tweaks

* about to test tutorial

* tutorial works

Corey-Zumar committed:
[CLIPPER-107][CLIPPER-109] Return JSON-formatted responses to prediction queries (#116)

* draft of the selection policy rewrite

* adapted query processor to use new selection policy interface

* minor cleanup

* added default output to application table

* things are compiling but there's a linker error from header-only json_utils

* unit tests passing

* addressed review comments and fixed end to end bench

* fixed unittests

* tweaks

* start json output

* json bool support, default output, return json

* Set object

* Error propagation

* Complete error handling, add tests

* Format code

* Remove unused and incorrect  method

* Address review comments

* Split exceptions into header and implementation

* Format code

* Fix tutorial, issue with PredictError

* format code

* Fix inconsistencies

* Fix rebase issues

* Fix unit tests, format code

* Remove unused imports, add newline

* Remove duplicate  field

* Revert container name change to clipper mgr

* Format code

* Fix frontend tests

* Add default output ratio

* rename ratio

Commit:
Merge branch 'get_app_info_api' of https://github.com/nishadsingh1/clipper into get_app_info_api
Moved implementaiton code in json_util.hpp to json_util.cpp

Commit:
Updated, refactored, and added tests

nishadsingh1 commented:
Made the recommended changes and:
- Moved function implementations out of `json_util.hpp` to `json_util.cpp`
- Renamed relevant function, most notably set_app_info_doc —> set_json_doc_from_redis_app_metadata. 
- set_json_doc_from_redis_app_metadata now makes use of several helper functions dedicated to transforming specific fields from their redis format to their publicly facing format
- Updated this transformation for the changed candidate model storage
- moved `get_all_applications` logic out of `redis.cpp`. Before, I had a function in `redis.cpp` iterating and making requests to redis. This seems like the sort of thing that shouldn’t exist in an adapter.

Added/updated the following tests:

**Management front-end tests**
TestGetApplicationCorrect
TestGetNonexistantApplicationCorrect
TestGetApplicationMalformedJson

TestGetAllApplicationsVerboseCorrect
TestGetAllApplicationsVerboseNoneRegisteredCorrect

TestGetAllApplicationsNotVerboseCorrect
TestGetAllApplicationsNotVerboseNoneRegisteredCorrect

TestGetAllApplicationsMalformedJson

**Redis tests**
ListApplicationNames
ListApplicationNamesNoneRegistered

**JSON util tests**
TestSetJsonDocFromRedisAppMetadata

Running the code formatting script made some changes in the files I touched. I included those changes here but I'm happy to remove them if the previous formatting was intentional.

nishadsingh1 commented:
I think that something about the way I rebased on develop and merged made this PR include a lot of changes that have already been made. I'm going to scrap it and make a new one – apologies for the confusion.
